cypress-io/cypress
# feat: Public API for CT Framework Definitions

Closes https://github.com/cypress-io/cypress/issues/25638

**Note**: No merge yet- still a known issue, see end of the PR description.

Code was reviewed separately in

- https://github.com/cypress-io/cypress/pull/25713
- https://github.com/cypress-io/cypress/pull/25749

A second pass is always good, though! 🙏 

- Technical Brief: https://github.com/cypress-io/engineering-documentation/blob/main/tech-briefs/ct-public-api/README.md
- Example third party Component Framework Definition: https://github.com/lmiller1990/cypress-ct-solid-js

## Screenshots

(1) it adds a community definition with relevant tag in launchpad

<img width="1165" alt="1" src="https://user-images.githubusercontent.com/19196536/218446285-3cf49015-0e4b-4c74-a5fa-01b1e28856dd.png">

(2) Known bug in dep detection - not introduced here - see #25782 

<img width="1165" alt="2" src="https://user-images.githubusercontent.com/19196536/218446378-837afd15-1b2b-4ac5-ab5e-fc142d240705.png">

(3) it works!

<img width="1512" alt="3" src="https://user-images.githubusercontent.com/19196536/218446707-8871d1dc-8457-4b8e-93a8-d0b1fe9e8903.png">


## Test Plan

1. make Solid.js app: `npx degit solidjs/templates/js my-app`
2. `cd my-app`
3. Install deps and my simple [Solid.js definition](https://github.com/lmiller1990/cypress-ct-solid-js): `npm install cypress-ct-lachlan-solid-js`

Two options, either
1. clone this branch, `yarn install` so you have the latest everything, `yarn cypress:open` at monorepo root for global mode, choose `my-app` (the one you just made)
2. grab a binary: https://github.com/cypress-io/cypress/commit/b2f6343ab7ae088c3907f38acdc10a73aafc517b#comments

Then...

1. Go through onboarding - Solid should show up. Use Vite, the template in step 1 is a Vite one.
2. It won't properly detect that you installed Solid.js. Known issue, unrelated to this PR, see #25782. Just hit Skip
3. Should work like any other CT lib, example test you can use:
4. wasn't that easy!! almost the same as a first party onboarding :D

```js
import App from '../../src/App'

describe('App', () => {
  it('playground', () => {
    cy.mount(<App />)
    cy.get('a').contains('Learn Solid')
  })
})
```

## Known issues/TODO before merge

- [x] ~Not working with binary. Definitions fail to resolve with `Ignoring TypeError: Invalid host defined options`. Seems to be V8 related according to [this tweet](https://twitter.com/slicknet/status/1559619176230047744), which is the only info I can find about this error. Working on it.~ Edit seems to be fixed https://github.com/cypress-io/cypress/commit/b2f6343ab7ae088c3907f38acdc10a73aafc517b#comments
- [ ] Fix flaky test (launchpad test for adding Qwik, a third party definition)

## Fast follows

- docs https://github.com/cypress-io/cypress-documentation/issues/5039
- #25782
- show at Cypress App demo
- GTM plan 

These are all out of scope for this PR - but we will do those next sprint.
